support mutations on global sets of constants.

laithsakka opened the pull request:
Disabling cpp guards to find unit tests that i need to update for now. 
this should only land on top of the cpp guards fix. and test_add_global_set should be updated
to run on cpp guards on.

Fix  https://github.com/pytorch/pytorch/issues/133509
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #133874



cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @chenyang78 @kadeng @chauhang @amjames

laithsakka committed:
support mutations on global sets

[ghstack-poisoned]

anijain2305 commented on torch/_dynamo/variables/builder.py:
```diff
@@ -560,11 +566,14 @@ class Autotuner:
                 # order, we just save the indices and use it to access keys and
                 # values. Indices are cheap to save.
+
+                # Do we need to guard on key order for sets?
+                # if not is_set:
```
No

anijain2305 commented on torch/_dynamo/variables/builder.py:
```diff
@@ -539,11 +539,17 @@ class Autotuner:
             result.source = self.source
             return self.tx.output.side_effects.track_object_existing(value, result)
-        elif istype(value, (dict, collections.defaultdict, collections.OrderedDict)):
+        elif istype(
+            value, (dict, collections.defaultdict, collections.OrderedDict, set)
+        ):
```
Just for readability, it would be good to have a separate if condition here.

anijain2305 commented on torch/_dynamo/variables/builder.py:
```diff
@@ -582,9 +591,23 @@ def build_key_value(i, k, v):
+                    source_key = ConstDictKeySource(self.get_source(), i)
+                    key = LazyVariableTracker.create(k, source_key)
+
+                source_value = GetItemSource(self.get_source(), source_key)
```
This is wrong. There is no such thing as `set[index]`.

laithsakka replied:
ah yeh no value is needed this line should be removed.

laithsakka commented on torch/_dynamo/variables/builder.py:
```diff
@@ -539,11 +539,17 @@ class Autotuner:
+
+            # Optimization for the common case strings, ints, etc
+            all_keys = value if is_set else value.keys()
+            all_const = all(ConstantVariable.is_literal(k) for k in all_keys)
```
GuardBuilder.DICT_CONST_KEYS would also need to be updated to support sets.

laithsakka committed:
Update on "[draft] support mutations on global sets"

A set in dynamo is implemented as a dictionary with None values, and set items as keys. 
Before this PR mutations on sets that are global was not supported, since mutable_local used to be none
to support it, I followed the path that dictionary takes, and since sets are just dictionaries I added a parallel 
support for sets.   

Fix  https://github.com/pytorch/pytorch/issues/133509



cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang amjames

[ghstack-poisoned]

laithsakka commented:
Re-implement after discussion with @anijain2305

laithsakka commented:
looking at faluires

laithsakka commented:
current approach does not correctly guard the test, debugging locally.

laithsakka committed:
Update on "support mutations on global sets"


Fix  https://github.com/pytorch/pytorch/issues/133509



cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang amjames

[ghstack-poisoned]

laithsakka committed:
Update on "support mutations on global sets"


Fix  https://github.com/pytorch/pytorch/issues/133509



cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang amjames

[ghstack-poisoned]

laithsakka committed:
Update on "support mutations on global sets"


Fix  https://github.com/pytorch/pytorch/issues/133509



cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang amjames

[ghstack-poisoned]

jansel commented on test/dynamo/test_functions.py:
```diff
@@ -3306,6 +3306,46 @@ def fn(x):
 
+    def test_add_global_set(self):
+        # This fail with enable_cpp_guard_manager = True.
+        torch._dynamo.config.enable_cpp_guard_manager = False
```
Use `config.patch` otherwise this will change the behavior for tests that run after thos one.

laithsakka replied:
i will remove this after rebase on https://github.com/pytorch/pytorch/pull/134016

jansel commented on torch/_dynamo/variables/builder.py:
```diff
@@ -539,10 +539,14 @@ class Autotuner:
             return self.tx.output.side_effects.track_object_existing(value, result)
+        elif istype(value, set):
+            self.install_guards(GuardBuilder.EQUALS_MATCH)
+            result = ConstantVariable.create(value=value)
```
How do we know the set is a constant?  Couldn't it contain non-constant objects?

laithsakka replied:
I will restrict this to when its literal

laithsakka committed:
Update on "support mutations on global sets"

Disabling cpp guards to find unit tests that i need to update for now. 
this should only land on top of the cpp guards fix. and test_add_global_set should be updated
to run on cpp guards on.

Fix  https://github.com/pytorch/pytorch/issues/133509



cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang amjames

[ghstack-poisoned]

laithsakka committed:
Update on "support mutations on global sets"

Disabling cpp guards to find unit tests that i need to update for now. 
this should only land on top of the cpp guards fix. and test_add_global_set should be updated
to run on cpp guards on.

Fix  https://github.com/pytorch/pytorch/issues/133509



cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang amjames

[ghstack-poisoned]

laithsakka committed:
Update on "support mutations on global sets"

Disabling cpp guards to find unit tests that i need to update for now. 
this should only land on top of the cpp guards fix. and test_add_global_set should be updated
to run on cpp guards on.

Fix  https://github.com/pytorch/pytorch/issues/133509



cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang amjames

[ghstack-poisoned]

laithsakka commented:
This will be abundant since the change is included in 
https://github.com/pytorch/pytorch/pull/134016/files
